src/core/storage/storage.types.ts

```diff
@@ -0,0 +1,139 @@
+import { v4 as uuidv4 } from "uuid";
+import { instanceToPlain, instanceToInstance } from "class-transformer";
+
+export type TagValue = string | boolean | undefined | Array<string> | null;
+export type TagsBase = {
+  [key: string]: TagValue;
+  [key: number]: never;
+};
+
+export type Tags<
+  DefaultTags extends TagsBase,
+  CustomTags extends TagsBase
+> = CustomTags & DefaultTags;
+
+export type RecordTags<Record extends BaseRecord> = ReturnType<
+  Record["getTags"]
+>;
+
+export type BaseRecordAny = BaseRecord<any, any>;
+
+abstract class BaseRecord<
+  DefaultTags extends TagsBase = TagsBase,
+  CustomTags extends TagsBase = TagsBase
+> {
+  protected _tags: CustomTags = {} as CustomTags;
+
+  public id!: string;
+
+  public createdAt!: Date;
+
+  public updatedAt?: Date;
+
+  public readonly type = BaseRecord.type;
+  public static readonly type: string = "BaseRecord";
+
+  public abstract getTags(): Tags<DefaultTags, CustomTags>;
+
+  public setTag(name: keyof CustomTags, value: CustomTags[keyof CustomTags]) {
+    this._tags[name] = value;
+  }
+
+  public getTag(name: keyof CustomTags | keyof DefaultTags) {
+    return this.getTags()[name];
+  }
+
+  public setTags(tags: Partial<CustomTags>) {
+    this._tags = {
+      ...this._tags,
+      ...tags,
+    };
+  }
+
+  public replaceTags(tags: CustomTags & Partial<DefaultTags>) {
+    this._tags = tags;
+  }
+
+  public toJSON(): Record<string, unknown> {
+    return instanceToPlain(this, {
+      exposeDefaultValues: true,
+    });
+  }
+
+  public clone() {
+    return instanceToInstance(this, {
+      exposeDefaultValues: true,
+      enableCircularCheck: true,
+      enableImplicitConversion: true,
+      ignoreDecorators: true,
+    });
+  }
+}
+
+type GenericRecordTags = TagsBase;
+interface BasicRecordStorageProps {
+  id?: string;
+  createdAt?: Date;
+  tags?: GenericRecordTags;
+  content: Record<string, unknown>;
+}
+interface SaveBasicRecordOption {
+  content: Record<string, unknown>;
+  id?: string;
+  tags?: GenericRecordTags;
+}
+
+class BasicRecord extends BaseRecord<GenericRecordTags> {
+  public content!: Record<string, unknown>;
+
+  public static readonly type = "GenericRecord";
+  public readonly type = BasicRecord.type;
```

Comment:
The type is used as the "category" at the schema layer for Askar etc - since this is "generic" can we actually pass the `type` as a mandatory prop? And have an enum of types we can use

Reply:
Good idea, I can try. It mean we can filter with this type like the GenericRecordType I defined before?

Reply:
Yep that will work. :) But lets also call it `enum RecordType`